pkg/networking/cilium/reconciler/reconciler.go

```diff
@@ -0,0 +1,183 @@
+package reconciler
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/controller"
+	"github.com/aws/eks-anywhere/pkg/controller/clientutil"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+	"github.com/aws/eks-anywhere/pkg/logger"
+	"github.com/aws/eks-anywhere/pkg/networking/cilium"
+	"github.com/aws/eks-anywhere/pkg/semver"
+	"github.com/aws/eks-anywhere/pkg/utils/oci"
+)
+
+const defaultRequeueTime = time.Second * 10
+
+type Templater interface {
+	GenerateUpgradePreflightManifest(ctx context.Context, spec *cluster.Spec) ([]byte, error)
+	GenerateManifest(ctx context.Context, spec *cluster.Spec, opts ...cilium.ManifestOpt) ([]byte, error)
+}
+
+// Reconciler allows to reconcile a Cilium CNI
+type Reconciler struct {
+	templater Templater
+}
+
+func New(templater Templater) *Reconciler {
+	return &Reconciler{
+		templater: templater,
+	}
+}
+
+// Reconcile takes the CIlium CNI in a cluster to the desired state defined in a cluster Spec
+// It uses a controller.Result to indicate when requeues are needed
+// Intended to be used in a kubernetes controller
+func (r *Reconciler) Reconcile(ctx context.Context, logger logr.Logger, client client.Client, spec *cluster.Spec) (controller.Result, error) {
+	installation, err := getInstallation(ctx, client)
```

Comment:
somewhat related but do we need a CNI health check to periodically monitor the status of Cilium and make sure it's all well and good? In my scaling testing, I encountered a few scenarios where Cilium had gone bad for a workload cluster and had to be reinstalled

Reply:
that should be done by default, since this will run in a controller-runtime controller: CRs get reconciled periodically (every x hours), even without change

That said, this won't reinstall if it already exists, it would just check its health. I think we should investigate why Cilium needed to be reinstalled. Any more details on what was broken and why it needed to be reinstalled?

Reply:
No more details. I just observed that after the upgrade, a few random workload clusters would be stuck with the machinedeployments not being ready and upon closer inspection, it showed that Cilium was down for those clusters. I never identified a root cause.

Reply:
Got you. Then I would propose to move forward with this code. If we see those problems again, we can dive deep and find the root cause. That should allow us to write the code here to detect such scenario and fix it (even if the fix is reinstalling).